ci: split and use matrix to test containerd backed image store

crazy-max opened the pull request:
follow-up https://github.com/moby/moby/pull/46402#pullrequestreview-1651587668

**- What I did**

Alternative to #46402 using a matrix and set required env vars to test containerd backed image store.

**- How I did it**

Sets a `storage` matrix conditionally based on `containerd-integration` GitHub label being set.

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

TBBle commented on .github/workflows/.windows.yml:
```diff
@@ -338,6 +364,11 @@ jobs:
           If ("${{ matrix.runtime }}" -eq "containerd") {
             $runtimeArg="--containerd=\\.\pipe\containerd-containerd"
             echo "DOCKER_WINDOWS_CONTAINERD_RUNTIME=1" | Out-File -FilePath $Env:GITHUB_ENV -Encoding utf-8 -Append
+            # FIXME: This doesn't really put the daemon on containerd snapshotter mode, see https://github.com/moby/moby/pull/46402#issuecomment-1710232310
+            If ("${{ matrix.storage }}" -eq "snapshotter") {
+              echo "TEST_INTEGRATION_USE_SNAPSHOTTER=1" | Out-File -FilePath $Env:GITHUB_ENV -Encoding utf-8 -Append
+              echo "DOCKER_GRAPHDRIVER=windows" | Out-File -FilePath $Env:GITHUB_ENV -Encoding utf-8 -Append
+            }
```
Fix is here: https://github.com/moby/moby/pull/46539/commits/c9c5594823a1c57a07734006ff41c1c7c5970274#diff-90e64dd59d8246305776c16fc6a5a006ef9458e5d0f7a6d64147f1c9ad97ddedL261-R283.

I removed `DOCKER_GRAPHDRIVER` as the Windows daemon doesn't honour it, and AFAIK doesn't plan to do so.

crazy-max replied:
Ok `TEST_INTEGRATION_USE_SNAPSHOTTER` is taken into account: https://github.com/moby/moby/actions/runs/6359507674/job/17272749812?pr=46572#step:21:72

```
Warning: 2023-09-30T02:28:08.8364201Z [Warning] Enabling containerd snapshotter through the $TEST_INTEGRATION_USE_SNAPSHOTTER environment variable. This should only be used for testing.
```

But looks like daemon fails to start: https://github.com/moby/moby/actions/runs/6359507674/job/17272749812?pr=46572#step:21:97

```
Error: 2023-09-30T02:28:08.9526619Z [Error] fatal: no host network support on this platform
```

crazy-max replied:
Ah I read your changes https://github.com/moby/moby/commit/c9c5594823a1c57a07734006ff41c1c7c5970274#diff-90e64dd59d8246305776c16fc6a5a006ef9458e5d0f7a6d64147f1c9ad97ddedL261-R283 wrong. Updated.

crazy-max replied:
Same fatal error: https://github.com/moby/moby/actions/runs/6359657942/job/17272998350?pr=46572#step:21:97

TBBle commented on .github/workflows/.windows.yml:
```diff
@@ -246,14 +269,17 @@ jobs:
           echo ${{ steps.tests.outputs.matrix }}
 
   integration-test:
+    if: ${{ matrix.runtime != 'builtin' && matrix.storage != 'snapshotter' }}
```
Should that be `||`? I'm guessing this is supposed to be excluding `snapshotter+builtin`.

Seems easier to add an exclude stanza to the strategy to me.

crazy-max commented on .github/workflows/.windows.yml:
```diff
@@ -249,15 +272,20 @@ jobs:
           - builtin
           - containerd
         test: ${{ fromJson(needs.integration-test-prepare.outputs.matrix) }}
+        exclude:
+          - storage: snapshotter
+            runtime: builtin
```
In the original PR, some jobs were dedup when running `builtin` along with `snapshotter`. This `exclude` makes sure we don't run them unnecessarily.

TBBle replied:
snapshotter+builtin is an impossible scenario, since the builtin (fake) libcontainerd implementation doesn't implement snapshotters. Same thing is needed in `integration-test-report`.

crazy-max replied:
> Same thing is needed in `integration-test-report`.

:+1:

crazy-max commented on .github/workflows/.windows.yml:
```diff
@@ -469,7 +502,7 @@ jobs:
         uses: actions/upload-artifact@v3
         with:
-          name: ${{ inputs.os }}-integration-reports-${{ matrix.runtime }}
+          name: ${{ inputs.os }}-${{ matrix.storage }}-integration-reports-${{ matrix.runtime }}
```
Was missing in the original PR as well.

thaJeztah commented:
just asking; does this approach allow for a separated status badge to be created for snapshotter?

basically if we want to enable this on master it'd be good if we could see that "regular" master is green (but it's fine to have the snapshotter variant fail for now)

crazy-max commented:
> just asking; does this approach allow for a separated status badge to be created for snapshotter?
> 
> basically if we want to enable this on master it'd be good if we could see that "regular" master is green (but it's fine to have the snapshotter variant fail for now)

No it's grouped under the same `test` workflow compared to original PR. Didn't catch we wanted also a split within the UI.

thaJeztah commented:
Yeah, the idea was that if we run it on master (so that we can always track the state of the containerd integration, and not just on PR's in that area), we don't want master to be "failing" app the time, so .. looking for a way to show master as "green" (but containerd integration still needing work)

crazy-max commented:
> Yeah, the idea was that if we run it on master (so that we can always track the state of the containerd integration, and not just on PR's in that area), we don't want master to be "failing" app the time, so .. looking for a way to show master as "green" (but containerd integration still needing work)

In this case we can use [`continue-on-error`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error) when built on a branch and not on PR so the run is green.

crazy-max commented:
Updated so now it splits based on storage input:

![image](https://github.com/moby/moby/assets/1951866/6b444682-329a-47e8-9c03-a332d0e329e0)

And added `continue-on-error` so it would not fail the workflow on non-PR if snapshotter is used.

crazy-max committed:
ci: continue-on-error on non-PR with snapshotter enabled

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

TBBle commented:
I _think_ the Windows artifacts all need to have their matrix tags added, as it looks like in the last run, the two build jobs might have stomped on each other's artifacts, leading to both the snapshotter and graphdriver integration test runs not finding _any_ build artifacts.

crazy-max commented:
> I _think_ the Windows artifacts all need to have their matrix tags added, as it looks like in the last run, the two build jobs might have stomped on each other's artifacts, leading to both the snapshotter and graphdriver integration test runs not finding _any_ build artifacts.
> 
> Alternatively, perhaps the failure cause is that the build job has been split across the image-store test split, but really should only happen once, and be consumed by both test streams. (Probably true for integration-test-prepare too, but I'm not sure it, or even the build job, can be unsplit like that, I didn't look that closely, just noticed the Windows graphdriver failures and assumed from there.,)
> 
> I did notice that the artefact `windows-2022-unit-reports` was generated twice, so I suspect _both_ things are true.

Thanks! I was not near my computer and just let the workflow run tbh :sweat_smile:
I'm taking a look :eyes:

rumpl commented on .github/workflows/windows-2019.yml:
```diff
@@ -13,10 +13,38 @@ jobs:
+            await core.group(`Set matrix`, async () => {
+              core.info(`matrix: ${JSON.stringify(matrix)}`);
+              core.setOutput('matrix', JSON.stringify(matrix));
+            });
```
Oh man

crazy-max commented:
Rebased in case you still want this.

rumpl commented:
> Rebased in case you still want this.

Oh we do want this

crazy-max committed:
ci: split tests in a reusable workflow

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
ci: use matrix to test containerd backed image store

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

rumpl reviewed (approved):
LGTM, my eyes are bleeding from so many yaml :D

laurazard reviewed (approved):
LGTM

thaJeztah reviewed (approved):
"reviewed" this PR after others already did

✅ it has spaces and indentation
✅ it's yaml
✅ CI runs

I'd say: LGTM let's get this in and we'll see if the world burns

thaJeztah commented:
^^ we should look at overlap in unit tests etc in a follow-up though; could someone create a ticket for that? (perhaps summarising parts to look into?)
